File: frontend/utils/api_client.py
import requests
from typing import Dict, List, Optional
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class APIClient:
    def __init__(self, base_url: str = "http://localhost:8000/api/v1"):
        self.base_url = base_url
        self.session = requests.Session()
    
    def create_post(self, content: str, author_alias: str = "Anonymous") -> Dict:
        """Create a new post"""
        try:
            response = self.session.post(
                f"{self.base_url}/posts",
                json={"content": content, "author_alias": author_alias},
                timeout=10
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            logger.error(f"Error creating post: {e}")
            raise Exception(f"Failed to create post: {str(e)}")
    
    def get_posts(self, page: int = 1, limit: int = 20) -> Dict:
        """Get paginated posts"""
        try:
            response = self.session.get(
                f"{self.base_url}/posts",
                params={"page": page, "limit": limit},
                timeout=10
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            logger.error(f"Error fetching posts: {e}")
            raise Exception(f"Failed to fetch posts: {str(e)}")
    
    def like_post(self, post_id: int) -> Dict:
        """Like a post"""
        try:
            response = self.session.post(
                f"{self.base_url}/posts/{post_id}/like",
                timeout=10
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            logger.error(f"Error liking post: {e}")
            raise Exception(f"Failed to like post: {str(e)}")
    
    def flag_post(self, post_id: int) -> Dict:
        """Flag a post"""
        try:
            response = self.session.post(
                f"{self.base_url}/posts/{post_id}/flag",
                timeout=10
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            logger.error(f"Error flagging post: {e}")
            raise Exception(f"Failed to flag post: {str(e)}")

File: frontend-react/src/services/apiClient.js
const API_BASE_URL = 'https://whisperwall-ultimate-project-backend.onrender.com/';

/**
 * API Client for WhisperWall Backend
 */
class APIClient {
  constructor(baseUrl = API_BASE_URL) {
    this.baseUrl = baseUrl;
  }

  /**
   * Make HTTP request with error handling
   */
  async request(endpoint, options = {}) {
    const url = `${this.baseUrl}${endpoint}`;
    
    try {
      const response = await fetch(url, {
        headers: {
          'Content-Type': 'application/json',
          ...options.headers,
        },
        ...options,
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.detail || `HTTP Error: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error(`API Error [${endpoint}]:`, error);
      throw error;
    }
  }

  /**
   * Get paginated posts
   * @param {number} page - Page number (default: 1)
   * @param {number} limit - Posts per page (default: 20)
   * @returns {Promise<{posts: Array, total: number, pages: number, has_prev: boolean, has_next: boolean}>}
   */
  async getPosts(page = 1, limit = 20) {
    const params = new URLSearchParams({ page, limit });
    return this.request(`/posts?${params}`);
  }

  /**
   * Create a new post
   * @param {string} content - Post content
   * @param {string} authorAlias - Author name (default: "Anonymous")
   * @returns {Promise<Object>} Created post object
   */
  async createPost(content, authorAlias = 'Anonymous') {
    return this.request('/posts', {
      method: 'POST',
      body: JSON.stringify({
        content,
        author_alias: authorAlias,
      }),
    });
  }

  /**
   * Like a post
   * @param {number} postId - ID of the post to like
   * @returns {Promise<Object>} Updated post object
   */
  async likePost(postId) {
    return this.request(`/posts/${postId}/like`, {
      method: 'POST',
    });
  }

  /**
   * Flag a post for moderation
   * @param {number} postId - ID of the post to flag
   * @returns {Promise<Object>} Updated post object
   */
  async flagPost(postId) {
    return this.request(`/posts/${postId}/flag`, {
      method: 'POST',
    });
  }
}

// Export singleton instance
const apiClient = new APIClient();
export default apiClient;
export { APIClient };
